### Objetivo: Encontrar regiões mais adensadas e maior renda per capita em São Paulo e outros municípios a fim de buscar maior quantidade de corridas premium como motorista de aplicativo.

A demanda por corridas premium em aplicativos como Uber e 99 pode ser influenciada por dois fatores principais: a densidade populacional e a renda per capita. É possível supor que, em áreas mais densamente povoadas, haja uma maior quantidade de pessoas se deslocando com frequência, o que tende a gerar mais solicitações de transporte por aplicativo. Nessas regiões, como em centros urbanos e bairros com alta concentração de habitantes, a lógica sugere que o uso desse tipo de serviço seja mais comum devido ao grande número de potenciais usuários.

Por outro lado, em locais onde a renda per capita é mais elevada, é razoável imaginar que o interesse por categorias premium, como Uber Black, seja maior. Isso se baseia na suposição de que pessoas com maior poder aquisitivo podem preferir serviços que ofereçam mais conforto e exclusividade, mesmo com tarifas mais altas. Assim, serviços que oferecem veículos de qualidade superior e motoristas com melhores qualificações podem atrair um público mais exigente, disposto a pagar por essa experiência.

Da mesma forma, categorias semi-premium, como Uber Comfort e 99 Plus, poderiam ser mais populares em regiões de alta renda, pois oferecem uma combinação de conforto a um custo relativamente mais acessível. Com base nisso, regiões de maior densidade populacional e renda mais alta poderiam ter uma demanda proporcionalmente maior por essas corridas premium e semi-premium, refletindo o perfil dos habitantes que vivem ou trabalham nessas localidades.

Em relação aos dados, as informações fornecidas pelo Censo serão essenciais para a descoberta das regiões mais adensadas e de maior renda per capita.

Hoje, nós estamos em 2024 e o censo de 2022 saiu preliminarmente contendo apenas os dados populacionais. No entanto, já que o estudo envolve população e renda, será utilizado os dados de 2010 que estão completos. Sabe-se que em 14 anos, teremos diferença nos dados, mas ao analisar o aumento populacional em São Paulo e São Bernardo do Campo, 1,7% e 5,9% respectivamente, de 2010 a 2022, provavelmente, não obtém-se grandes diferenças habitacionais no período. No entanto, a atualização deste estudo se torna peremptória para a assertividade do objetivo estipulado assim que divulgado os dados do Censo de 2022.

O que são os setores censitários?

"Os setores censitários são a menor porção de área utilizadas pelo IBGE para planejar, coletar e disseminar os resultados dos Censos e Pesquisas Estatísticas. 
        Esses setores são definidos a partir de diversos critérios:
        * A área e o número de domicílios do setor devem ser passíveis de serem percorridos e visitados por um único recenseador em tempo hábil.
        * O recorte do setor deve ser representativo em relação à população local.
        * O setor deve ter um número de domicílios suficiente para que seus dados agregados preservem o sigilo dos moradores"
        (G1 - https://g1.globo.com/economia/censo/noticia/2024/03/21/o-que-e-o-setor-censitario-medida-usada-pelo-ibge-para-fazer-o-censo.ghtml)

Dados:
    
  Diretório do censo de todos os estados do Brasil por setor censitário:
    https://ftp.ibge.gov.br/Censos/Censo_Demografico_2010/Resultados_do_Universo/Agregados_por_Setores_Censitarios/
    
  Diretório dos arquivos geográficos dos setores censitários:
    https://geoftp.ibge.gov.br/organizacao_do_territorio/malhas_territoriais/malhas_de_setores_censitarios__divisoes_intramunicipais/censo_2010/setores_censitarios_shp/

A seguir será necessário estratificar a população de uma forma a relacionar a renda e densidade demográfica. Portanto, pesquisou-se artigos e informações relevantes para determinar a estratificação apresentados a seguir.

O primeirto artigo "NÍVEIS DE DENSIDADE POPULACIONAL: Uma proposta de classificação para Goiânia – GO, aplicação no Setor Central" (https://xviiienanpur.anpur.org.br/anaisadmin/capapdf.php?reqid=1034) apresenta as seguintes densidades populacionais como, em habitantes/hectare, " <15 muito baixa, de 16 a 50 baixa, de 51 a 150 média, de 151 a 350 alta e > 350 muito alta."

Além disso, utilizando a informação do artigo do site einvestidor.estadao.com.br ("Quanto ganha a classe A no Brasil?" Quanto ganha a classe A no Brasil? https://einvestidor.estadao.com.br/radar-einvestidor/quanto-ganha-a-classe-a-no-brasil/) obtém-se os dados referentes as rendas domiciliares por classe social:
- Classe A: R$ 28.240;
- Classe B1: R$ 12.683,34;
- Classe B2: R$ 7.017,64;
- Classe C1: R$ 3.980,38;
- Classe C2: R$ 2.403,04;
- Classe D-E: R$ 1.087,77.

Ademais, o artigo Tendências e Desigualdades da Mobilidade Urbana no Brasil II: Características e Padrões de Consumo da Mobilidade por Aplicativo”, conduzido por Lucas Warwar e Rafael H. M. Pereira (https://repositorio.ipea.gov.br/bitstream/11058/11240/1/td_2781.pdf) demonstra que "A distribuição dos usuários de ride-hailing (viagens por aplicativos) por faixas de renda evidencia uma forte concentração entre os estratos superiores".

De acordo com informações anteriores, consegue-se obter os dados relevantes das condições sócio-demográficas para a descoberta das regiões mais propensas a se obter viagens por aplicativo. Portanto, os dados deverão ser estratificados em regiões com densidade demográfica maior que 150 hab/ha e renda domiciliar maior ou igual a R$ 7.017,64.

---

## Importação de bibliotecas

In [1]:
import os # Criação de estrutura de pastas e manuseamento de arquivos
from ftplib import FTP # Biblioteca para download dos arquivos de servidor ftp
from zipfile import ZipFile # Biblioteca para descompactar os arquivos
import pandas as pd # Tratamento dos dados 
import geopandas as gpd # Tratamento dos dados geográficos
import matplotlib.pyplot as plt # Biblioteca para visualização gráfica dos dados
import folium # Biblioteca para visualização em mapas

---

## Estrutura de pastas

In [2]:
parent_dir = os.getcwd() # Identificação do diretório de trabalho

"""
Criação dos diretórios para arquivamento dos arquivos
data: Arquivos baixados e a serem utilizados para tratamento.
artifacts: Todos os arquivos criados para utilização e visualização.
"""

directories = ["data", "artifacts"] 
for directory in directories:
    path = os.path.join(parent_dir, directory)
    try:
        os.mkdir(path)
    except FileExistsError:
        print("Directory already created!")
        
data_dir = os.path.join(parent_dir, directories[0])
artifacts_dir = os.path.join(parent_dir, directories[1])

Directory already created!
Directory already created!


### Download de arquivos

Acessando os servidores FTP e baixando os arquivos.

In [3]:
ftp_server = "ftp.ibge.gov.br" # Servidor FTP do IBGE
ftp_dir = "Censos/Censo_Demografico_2010/Resultados_do_Universo/"\
    "Agregados_por_Setores_Censitarios/" # Diretório a ser acessado
# O servidor aceita acesso público, portanto user e password desnecessários
user = "anonymous" # Inserir anonymous para entrar no servidor
passwd = ""

ftp = FTP(ftp_server, user, passwd) # Configuração da variável de acesso ao servidor
ftp.cwd(ftp_dir) # Acessando o diretório escolhido
files = ftp.nlst() # variável contendo o conteúdo do diretório

ftp.dir() # Listagem do conteúdo do diretório

-rwxrwxr-x    1 ftp      ftp          1979 Oct 30  2023 1_Atualizacoes_20231030.txt
-rwxrwxr-x    1 ftp      ftp       5579523 Oct 30  2023 AC_20231030.zip
-rwxrwxr-x    1 ftp      ftp      22399914 Oct 30  2023 AL_20231030.zip
-rwxrwxr-x    1 ftp      ftp      31064700 Oct 30  2023 AM_20231030.zip
-rwxrwxr-x    1 ftp      ftp       5194028 Oct 30  2023 AP_20231030.zip
-rwxrwxr-x    1 ftp      ftp      126886126 Oct 30  2023 BA_20231030.zip
-rwxrwxr-x    1 ftp      ftp      71781739 Oct 30  2023 CE_20231030.zip
-rwxrwxr-x    1 ftp      ftp      21585996 Oct 30  2023 DF_20231030.zip
-rwxrwxr-x    1 ftp      ftp      34133643 Oct 30  2023 Documentacao_Agregado_dos_Setores_2010_20231030.zip
-rwxrwxr-x    1 ftp      ftp      37084676 Oct 30  2023 ES_20231030.zip
-rwxrwxr-x    1 ftp      ftp      53664403 Oct 30  2023 GO_20231030.zip
-rwxrwxr-x    1 ftp      ftp      48421505 Oct 30  2023 MA_20231030.zip
-rwxrwxr-x    1 ftp      ftp      169948641 Oct 30  2023 MG_20231030.zip
-rwxrwxr-x    

Nota-se que temos 3 tipos de arquivos:
  * 1 arquivo texto
  * 1 arquivo de documentação compactado
  * Arquivos do censo de cada estado compactado

In [4]:
docum = "Documentacao_Agregado" # variável para obter o arquivo de documentação
UFs = ["SP"] # lsita dos estados desejados
files_check = [ filename for filename in os.listdir(data_dir) ] # lista dos arquivos no diretório 

# Loop para download dos arquivos no servidor FTP
for file in files:
        if (file.startswith(docum) or file[:2] in UFs) and (file not in files_check):
            with open(os.path.join(data_dir,file), 'wb') as fp:
                ftp.retrbinary('RETR ' + file, fp.write)
                print(file+" downloaded to "+data_dir)

ftp.quit() # fechamento da conexão com o servidor FTP

'221 Goodbye.'

In [5]:
# Acessando o servidor ftp com as informações geográficas dos setores censitários
ftp_server = "geoftp.ibge.gov.br"
ftp_dir = "organizacao_do_territorio/malhas_territoriais/"\
    "malhas_de_setores_censitarios__divisoes_intramunicipais/censo_2010/"\
    "setores_censitarios_shp/"
user = "anonymous"
passwd = ""

ftp = FTP(ftp_server, user, passwd)
ftp.cwd(ftp_dir)
ftp.dir()

drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 1_leia_me
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 ac
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 al
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 am
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 ap
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 ba
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 ce
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 df
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 es
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 go
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 ma
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 mg
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 ms
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 mt
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 pa
drwxrwxr-x    2 ftp      ftp          4096 Jun 14  2016 pb
drwxrwxr-x    2 ftp      ftp          4096 Jun 14

In [6]:
ftp.dir('sp')

-rw-rw-r--    1 ftp      ftp      11116004 Jun 14  2016 sp_distritos.zip
-rw-rw-r--    1 ftp      ftp       9387313 Jun 14  2016 sp_municipios.zip
-rwxrwxr-x    1 ftp      ftp      45523805 Nov 04  2020 sp_setores_censitarios.zip
-rw-rw-r--    1 ftp      ftp      11108462 Jun 14  2016 sp_subdistritos.zip


Verifica-se que dentro do diretório há quatro arquivos com as divisões dos municípios, distritos, subdistritos e setores censitários. Neste momento, deseja-se os setores censitários.

In [7]:
UFs_geo = ['sp'] # lista dos estados
extension_geo = "_setores_censitarios.zip" # variável com a extensão dos arquivos a serem baixados
files_check = [ filename for filename in os.listdir(data_dir) ] 

# Loop entre todos os diretórios dos estados para download dos arquivos geográficos
for UF_geo in UFs_geo:
    ftp.cwd(UF_geo) 
    files = ftp.nlst()
    for file in files:
        if (file == UF_geo+extension_geo) and (file not in files_check):
            with open(os.path.join(data_dir,file), 'wb') as fp:
                ftp.retrbinary('RETR ' + file, fp.write)
                print(file+" downloaded to "+data_dir)
    ftp.cwd("..") # Voltando ao diretório anterior para repetição do processo
ftp.quit()

'221 Goodbye.'

---

## Extração dos arquivos compactados.

In [8]:
data_dir_files = os.listdir(data_dir) # Lista dos arquivos no diretório 'data'
data_dir_files

['Documentacao_Agregado_dos_Setores_2010_20231030',
 'Documentacao_Agregado_dos_Setores_2010_20231030.zip',
 'SP_Capital_20231030',
 'SP_Capital_20231030.zip',
 'SP_Exceto_Capital_20231030',
 'SP_Exceto_Capital_20231030.zip',
 'sp_setores_censitarios',
 'sp_setores_censitarios.zip']

In [9]:
# Loop para extração dos arquivos compactados
for file in data_dir_files:
    try:
        if file.split(".")[1] == "zip": # Verificação se são arquivos .zip
            file_zipped = ZipFile(os.path.join(data_dir,file))
            
            new_folder_path = os.path.join(data_dir, file.split(".zip")[0]) # Diretório para receber os arquivos
            
            # Criação do diretório caso não exista
            if not os.path.exists(new_folder_path):
                file_zipped.extractall(new_folder_path)
            
            file_zipped.close()
    except:
        print(f"{file} - Not a zipped file")

Documentacao_Agregado_dos_Setores_2010_20231030 - Not a zipped file
SP_Capital_20231030 - Not a zipped file
SP_Exceto_Capital_20231030 - Not a zipped file
sp_setores_censitarios - Not a zipped file


In [10]:
files_list =[]

for dirpath, dirnames, filenames in os.walk(data_dir):
    for filename in filenames:
        file_path = os.path.join(dirpath, filename)
        files_list.append(file_path)
        
files_list

['c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\Documentacao_Agregado_dos_Setores_2010_20231030.zip',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\SP_Capital_20231030.zip',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\SP_Exceto_Capital_20231030.zip',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\sp_setores_censitarios.zip',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\Documentacao_Agregado_dos_Setores_2010_20231030\\Documentaç╞o\\BASE DE INFORMAÇσES POR SETOR CENSIT╡RIO Censo 2010 - Universo novo.pdf',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\Documentacao_Agregado_dos_Setores_2010_20231030\\Documentaç╞o\\Descriç╞o_AC.xls',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\Documentacao_Agregado_dos_Setores_2010_20231030\\Documentaç╞o\\Descriç╞o_AL.xls',
 'c:\\Users\\guicr\\OneDrive\\Carreira\\Portfolio\\uber\\data\\Documentacao_Agregado_dos_Setores_2010_20231030\\Docu

Ao observar a lista de arquivos, percebe-se uma duplicidade entre os arquivos .csv e .xls referentes aos dados dos setores censitários. Então, podemos eliminar os arquivos duplicados que estão nas pastas com o formato: c:/Users/guicr/OneDrive/Carreira/Portfolio/uber/data/SP_Capital_20231030/Base informaçoes setores2010 universo SP_Capital/EXCEL/

In [11]:
# Diretórios a serem excluídos
mask = [os.path.join(data_dir, "SP_Capital_20231030\\Base informaçoes setores2010 universo SP_Capital\\EXCEL"),
        os.path.join(data_dir, "SP_Exceto_Capital_20231030\\Base informaçoes setores2010 universo SP_Exceto_Capital\\EXCEL")
        ]

# Loop para remoção dos arquivos e diretório
for item in mask:
    for dirpath, dirnames, filenames in os.walk(item, topdown=False):
        for filename in filenames:
                os.remove(os.path.join(dirpath, filename))
        os.rmdir(dirpath)

---

## Tratamento de dados

In [12]:
# Passando a lista de arquivos para um dataframe para utilização mais fácil dos arquivos
files_list_df =pd.DataFrame(files_list, columns=['filepaths'])
files_list_df

filepaths
0   c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
1   c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
2   c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
3   c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
4   c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
..                                                ...
84  c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
85  c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
86  c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
87  c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...
88  c:\Users\guicr\OneDrive\Carreira\Portfolio\ube...

[89 rows x 1 columns]

Agora, será necessário utilizar os arquivos necessários do censo para retirar a população e renda domiciliar de cada setor censitário, para isso serão utilizados dois arquivos de acordo com a documentação e eles possuem os seguintes formatos Basico_UF.csv, Domicilio01_UF e DomicilioRenda_UF(necessário trocar UF pelo estado a ser observado). No caso de São Paulo, existem dois tipos de arquivo um para a capital e outro para os demais municípios, os quais unir-se-ão adiante. Além disso, a renda por domicílio será obtida através da divisão da renda total por domicílio e a quantidade de domicílios fornecidos pelos arquivos DomicilioRenda_UF e Domicilio01_UF, respectivamente.

In [13]:
# Nome dos arquivos para capital
file_str_earning_total_dom_capital = 'DomicilioRenda_' + 'SP1' # Total de domicílios
file_str_total_dom_capital =  'Domicilio01_' + 'SP1' # Total da renda dos domicílios
file_str_pop_cap = 'pessoa13_' + 'sp1' # população

# Nome dos arquivos para os demais municípios
file_str_earning_total_dom = 'DomicilioRenda_' + 'SP2'
file_str_total_dom =  'Domicilio01_' + 'SP2'
file_str_pop = 'pessoa13_' + 'sp2'

In [14]:
# Obtendo os caminhos para os arquivos

# Filtros para os caminhos 
get_path_earning_total_dom_capital = files_list_df['filepaths'].str.contains(file_str_earning_total_dom_capital)
get_path_total_dom_capital = files_list_df['filepaths'].str.contains(file_str_total_dom_capital)
get_path_pop_capital = files_list_df['filepaths'].str.contains(file_str_pop_cap)

get_path_earning_total_dom = files_list_df['filepaths'].str.contains(file_str_earning_total_dom)
get_path_total_dom = files_list_df['filepaths'].str.contains(file_str_total_dom)
get_path_pop = files_list_df['filepaths'].str.contains(file_str_pop)

# Caminhos para renda domiciliar e quantidade de domicílios
file_path_earning_total_dom_capital = files_list_df.loc[get_path_earning_total_dom_capital, 'filepaths'].values[0]
file_path_total_dom_capital = files_list_df.loc[get_path_total_dom_capital, 'filepaths'].values[0]
file_path_earning_total_dom = files_list_df.loc[get_path_earning_total_dom, 'filepaths'].values[0]
file_path_total_dom = files_list_df.loc[get_path_total_dom, 'filepaths'].values[0]
# Caminhos para população
file_path_pop_capital = files_list_df.loc[get_path_pop_capital,'filepaths'].values[0]
file_path_pop = files_list_df.loc[get_path_pop,'filepaths'].values[0]

print(file_path_earning_total_dom_capital)
print(file_path_total_dom_capital)
print(file_path_pop_capital)
print(file_path_earning_total_dom)
print(file_path_total_dom)
print(file_path_pop)

c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\SP_Capital_20231030\Base informaçoes setores2010 universo SP_Capital\CSV\DomicilioRenda_SP1.csv
c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\SP_Capital_20231030\Base informaçoes setores2010 universo SP_Capital\CSV\Domicilio01_SP1.csv
c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\SP_Capital_20231030\Base informaçoes setores2010 universo SP_Capital\CSV\pessoa13_sp1.csv
c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\SP_Exceto_Capital_20231030\Base informaçoes setores2010 universo SP_Exceto_Capital\CSV\DomicilioRenda_SP2.csv
c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\SP_Exceto_Capital_20231030\Base informaçoes setores2010 universo SP_Exceto_Capital\CSV\Domicilio01_SP2.csv
c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\SP_Exceto_Capital_20231030\Base informaçoes setores2010 universo SP_Exceto_Capital\CSV\pessoa13_sp2.csv


In [15]:
# Passando os arquivos sobre domicílios .csv para dataframe
city_earning_cap = pd.read_csv(file_path_earning_total_dom_capital, encoding='iso-8859-1', sep=';')
city_dom_cap =  pd.read_csv(file_path_total_dom_capital, encoding='iso-8859-1', sep=';')
city_earning = pd.read_csv(file_path_earning_total_dom, encoding='iso-8859-1', sep=';')
city_dom =  pd.read_csv(file_path_total_dom, encoding='iso-8859-1', sep=';')

#### Executando algumas verificações nos arquivos...

    Verificação dos dados referentes a renda domiciliar:

In [16]:
# Observando o dataframe
city_earning_cap.head().T

0                1                2  \
Cod_setor       355030801000001  355030801000002  355030801000003   
Situacao_setor                1                1                1   
V001                          2                0                0   
V002                     907777           846525           505662   
V003                     903817           846525           505662   
V004                       3960                0                0   
V005                          0                0                0   
V006                          0                2                3   
V007                         14               28               11   
V008                         51               87               38   
V009                         94               94               50   
V010                         49               45               19   
V011                         42               30               16   
V012                         12               12               10   
V013                          5                4                3   
V014                          1                4               39   
Unnamed: 16                 NaN              NaN              NaN   

                              3                4  
Cod_setor       355030801000004  355030801000005  
Situacao_setor                1                1  
V001                          0                0  
V002                     446011           615215  
V003                     446011           615215  
V004                          0                0  
V005                          0                0  
V006                          3                2  
V007                          8               20  
V008                         39               68  
V009                         57               76  
V010                         14               37  
V011                         14               18  
V012                          9                8  
V013                          1                3  
V014                         36                8  
Unnamed: 16                 NaN              NaN

In [17]:
city_earning.head().T

0                1                2  \
Cod_setor       350010505000001  350010505000002  350010505000003   
Situacao_setor                1                1                1   
V001                          0                0                1   
V002                     439255           866389           921529   
V003                     439255           866389           917329   
V004                          0                0             4200   
V005                          0                0                0   
V006                          0                0                0   
V007                          1               17                1   
V008                         23               67               31   
V009                         43              100               56   
V010                         28               35               41   
V011                         25               41               40   
V012                         22               27               34   
V013                          4                3               10   
V014                          3                4                2   
Unnamed: 16                 NaN              NaN              NaN   

                              3                4  
Cod_setor       350010505000004  350010505000005  
Situacao_setor                1                1  
V001                          0                0  
V002                    1160677          1074174  
V003                    1160677          1074174  
V004                          0                0  
V005                          0                0  
V006                          1                1  
V007                         13               18  
V008                         56               66  
V009                         80               98  
V010                         38               45  
V011                         43               34  
V012                         42               21  
V013                         14                6  
V014                          1                2  
Unnamed: 16                 NaN              NaN

In [18]:
# Informações sobre as colunas
city_earning_cap.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18363 entries, 0 to 18362
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Cod_setor       18363 non-null  int64  
 1   Situacao_setor  18363 non-null  int64  
 2   V001            18363 non-null  object 
 3   V002            18363 non-null  object 
 4   V003            18363 non-null  object 
 5   V004            18363 non-null  object 
 6   V005            18363 non-null  object 
 7   V006            18363 non-null  object 
 8   V007            18363 non-null  object 
 9   V008            18363 non-null  object 
 10  V009            18363 non-null  object 
 11  V010            18363 non-null  object 
 12  V011            18363 non-null  object 
 13  V012            18363 non-null  object 
 14  V013            18363 non-null  object 
 15  V014            18363 non-null  object 
 16  Unnamed: 16     0 non-null      float64
dtypes: float64(1), int64(2), object

In [19]:
city_earning.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47733 entries, 0 to 47732
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Cod_setor       47733 non-null  int64  
 1   Situacao_setor  47733 non-null  int64  
 2   V001            47733 non-null  object 
 3   V002            47733 non-null  object 
 4   V003            47733 non-null  object 
 5   V004            47733 non-null  object 
 6   V005            47733 non-null  object 
 7   V006            47733 non-null  object 
 8   V007            47733 non-null  object 
 9   V008            47733 non-null  object 
 10  V009            47733 non-null  object 
 11  V010            47733 non-null  object 
 12  V011            47733 non-null  object 
 13  V012            47733 non-null  object 
 14  V013            47733 non-null  object 
 15  V014            47733 non-null  object 
 16  Unnamed: 16     0 non-null      float64
dtypes: float64(1), int64(2), object

Em ambos os casos, temos 16 colunas com uma tipo float, duas tipo int e 14 tipo object. Temos muitas colunas nesses dois dataframes, mas, no caso de renda domiciliar, necessitamos apenas do código do setor censitário e o total do rendimento nominal mensal dos domicílios particulares informados, segundo o documentação, pelas colunas Cod_setor e V002.

Quais são as colunas dos dataframes anteriores?

In [20]:
list(city_earning_cap.columns)

['Cod_setor',
 'Situacao_setor',
 'V001',
 'V002',
 'V003',
 'V004',
 'V005',
 'V006',
 'V007',
 'V008',
 'V009',
 'V010',
 'V011',
 'V012',
 'V013',
 'V014',
 'Unnamed: 16']

In [21]:
list(city_earning.columns)

['Cod_setor',
 'Situacao_setor',
 'V001',
 'V002',
 'V003',
 'V004',
 'V005',
 'V006',
 'V007',
 'V008',
 'V009',
 'V010',
 'V011',
 'V012',
 'V013',
 'V014',
 'Unnamed: 16']

Criação de dois dataframes contendo apenas as informações relevantes.

In [22]:
stg_city_earning_cap = city_earning_cap.loc[:,['Cod_setor', 'V002']]
stg_city_earning = city_earning.loc[:,['Cod_setor', 'V002']]

In [23]:
stg_city_earning.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47733 entries, 0 to 47732
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Cod_setor  47733 non-null  int64 
 1   V002       47733 non-null  object
dtypes: int64(1), object(1)
memory usage: 746.0+ KB


In [24]:
stg_city_earning_cap.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18363 entries, 0 to 18362
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Cod_setor  18363 non-null  int64 
 1   V002       18363 non-null  object
dtypes: int64(1), object(1)
memory usage: 287.0+ KB


In [25]:
stg_city_earning.head()

Cod_setor     V002
0  350010505000001   439255
1  350010505000002   866389
2  350010505000003   921529
3  350010505000004  1160677
4  350010505000005  1074174

In [26]:
stg_city_earning_cap.head()

Cod_setor    V002
0  355030801000001  907777
1  355030801000002  846525
2  355030801000003  505662
3  355030801000004  446011
4  355030801000005  615215

Nota-se que o tipo de dado da renda é object e código do setor censitário é int. No entanto, o a renda deveria ser do tipo float e o código do setor censitário tipo str, pois trata-se de uma identificação que utiliza-se para agregar os dados mais à frente. Por isso, trocam-se dos tipos dos dados.

In [27]:
stg_city_earning_cap = stg_city_earning_cap.astype({'Cod_setor' : 'str',
                                                    'V002' : 'float'})
stg_city_earning = stg_city_earning.astype({'Cod_setor' : 'str',
                                                    'V002' : 'float'})

ValueError: could not convert string to float: 'X': Error while type casting for column 'V002'

Existe um dado na coluna V002 que não é numérico, verifica-se se existem mais possibilidades.

In [28]:
mask = pd.to_numeric(stg_city_earning['V002'], errors='coerce').isna()
non_numeric_values = stg_city_earning[mask]
print(non_numeric_values)

             Cod_setor V002
31     350010505000032    X
43     350010505000045    X
52     350010505000056    X
62     350010505000068    X
64     350010505000070    X
...                ...  ...
47442  355680005000026    X
47448  355680005000032    X
47449  355680005000033    X
47696  355715405000003    X
47698  355715405000005    X

[1222 rows x 2 columns]


In [29]:
mask = pd.to_numeric(stg_city_earning_cap['V002'], errors='coerce').isna()
non_numeric_values = stg_city_earning_cap[mask]
print(non_numeric_values)

             Cod_setor V002
241    355030803000034    X
250    355030803000043    X
251    355030803000044    X
272    355030803000071    X
283    355030803000085    X
...                ...  ...
17541  355030892000188    X
17881  355030894000153    X
17991  355030894000268    X
18079  355030895000087    X
18101  355030895000110    X

[157 rows x 2 columns]


Percebe-se que existem valores marcados com 'X', esses valores faltantes estão relacionados com a política de não divulgação de dados adotado pelo IBGE descrito na documentação. Portanto, criam-se novos dataframes sem estes dados para poder mudar o tipos dos dados.

In [31]:
stg1_city_earning = stg_city_earning[~pd.to_numeric(stg_city_earning['V002'], errors='coerce').isna()].copy()
stg1_city_earning_cap = stg_city_earning_cap[~pd.to_numeric(stg_city_earning_cap['V002'], errors='coerce').isna()].copy()
stg1_city_earning_cap = stg1_city_earning_cap.astype({'Cod_setor' : 'str',
                                                    'V002' : 'float'})
stg1_city_earning = stg1_city_earning.astype({'Cod_setor' : 'str',
                                                    'V002' : 'float'})

In [32]:
stg1_city_earning_cap.head()

Cod_setor      V002
0  355030801000001  907777.0
1  355030801000002  846525.0
2  355030801000003  505662.0
3  355030801000004  446011.0
4  355030801000005  615215.0

In [34]:
stg1_city_earning.head()

Cod_setor       V002
0  350010505000001   439255.0
1  350010505000002   866389.0
2  350010505000003   921529.0
3  350010505000004  1160677.0
4  350010505000005  1074174.0

In [35]:
stg1_city_earning.dtypes

Cod_setor     object
V002         float64
dtype: object

In [36]:
stg1_city_earning_cap.dtypes

Cod_setor     object
V002         float64
dtype: object

In [37]:
# Quantidade de dados nulos por coluna
stg1_city_earning_cap.isnull().sum()

Cod_setor    0
V002         0
dtype: int64

In [38]:
stg1_city_earning.isnull().sum()

Cod_setor    0
V002         0
dtype: int64

Tudo corrigido e formatado para os próximos passos.

    Verificação dos dados referente aos domicílios

In [39]:
stg_city_dom_cap = city_dom_cap.copy()
stg_city_dom = city_dom.copy()

In [40]:
stg_city_dom.head()

Cod_setor  Situacao_setor  V001  V002 V003 V004 V005 V006 V007 V008  \
0  350010505000001               1   149   149   83    0   58   67    0   70   
1  350010505000002               1   294   294  287    0    7  185    4   88   
2  350010505000003               1   219   214  188    0   26  156    2   37   
3  350010505000004               1   295   288  284    0    4  187    7   78   
4  350010505000005               1   291   291  274    1   16  211    1   62   

   ... V233 V234 V235 V236 V237 V238 V239 V240 V241 Unnamed: 243  
0  ...    0    0    0    0    0    0    0    0    0          NaN  
1  ...    0    1    1    0    0    0    0    0    0          NaN  
2  ...    0    0    0    0    0    0    0    0    0          NaN  
3  ...    0    1    1    0    0    0    0    0    0          NaN  
4  ...    0    0    0    0    0    0    0    0    0          NaN  

[5 rows x 244 columns]

Os dados referentes aos domicílios possuem 244 colunas. Portanto, reduz-se a dimensionalidade apenas para os dados relevantes para esse estudo são 'Cod_setor' e 'V001'

In [49]:
stg1_city_dom = stg_city_dom[['Cod_setor', 'V001']].copy()
stg1_city_dom_cap = stg_city_dom_cap[['Cod_setor', 'V001']].copy()

In [50]:
stg1_city_dom.head()

Cod_setor  V001
0  350010505000001   149
1  350010505000002   294
2  350010505000003   219
3  350010505000004   295
4  350010505000005   291

In [51]:
stg1_city_dom_cap.head()

Cod_setor  V001
0  355030801000001   268
1  355030801000002   306
2  355030801000003   189
3  355030801000004   181
4  355030801000005   240

In [52]:
stg1_city_dom_cap = stg1_city_dom_cap.astype({'Cod_setor' : 'str',
                                                    'V001' : 'int'})
stg1_city_dom = stg1_city_dom.astype({'Cod_setor' : 'str',
                                                    'V001' : 'int'})

Necessita-se unir os dataframes para consolidação dos dados.

Primeiro, concatena-se os dados dos municípios e da capital.

In [59]:
comb_earning_city = pd.concat([stg1_city_earning_cap, stg1_city_earning])
comb_dom_city = pd.concat([stg1_city_dom_cap, stg1_city_dom])

In [64]:
print(comb_earning_city.info())
print(comb_dom_city.info())

<class 'pandas.core.frame.DataFrame'>
Index: 64717 entries, 0 to 47732
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Cod_setor  64717 non-null  object 
 1   V002       64717 non-null  float64
dtypes: float64(1), object(1)
memory usage: 1.5+ MB
None
<class 'pandas.core.frame.DataFrame'>
Index: 66096 entries, 0 to 47732
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Cod_setor  66096 non-null  object
 1   V001       66096 non-null  int32 
dtypes: int32(1), object(1)
memory usage: 1.3+ MB
None


Analisa-se as informações dos dados e percebe-se que a quantidade dos dados diferem em ambas as situações renda domiciliar e domicílios. Portanto, manteve-se os dados com menor quantidade de dimensões.

In [66]:
comb_city = pd.merge(comb_earning_city, comb_dom_city,how='left', on='Cod_setor')

In [67]:
comb_city.head()

Cod_setor      V002  V001
0  355030801000001  907777.0   268
1  355030801000002  846525.0   306
2  355030801000003  505662.0   189
3  355030801000004  446011.0   181
4  355030801000005  615215.0   240

In [70]:
comb_city.rename(columns={'V002':'renda',
                          'V001':'domicilios'},
                 inplace=True)

In [71]:
comb_city.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64717 entries, 0 to 64716
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Cod_setor   64717 non-null  object 
 1   renda       64717 non-null  float64
 2   domicilios  64717 non-null  int32  
dtypes: float64(1), int32(1), object(1)
memory usage: 1.2+ MB


In [72]:
comb_earning_city.head()

Cod_setor      V002
0  355030801000001  907777.0
1  355030801000002  846525.0
2  355030801000003  505662.0
3  355030801000004  446011.0
4  355030801000005  615215.0

In [74]:
comb_city.isnull().sum()

Cod_setor     0
renda         0
domicilios    0
dtype: int64

Agora, tem-se um único dataframe com as informações relevantes dos domicílios e colunas com nomes mais explicativos.

---

## Tratamento dos dados geográficos

Nesta situação, os dados provenientes do arquivo shapefile(.shp) informarão os dados geográficos para localização dos setores e cálculos de área de cada setor censitário. O arquivo para o estado de São Paulo é o 35SEE250GC_SIR.shp.

In [75]:
# Localizando o arquivo
file_str_geo = '35SEE250GC_SIR.shp'
get_path_geo = files_list_df['filepaths'].str.contains(file_str_geo)
file_path_geo = files_list_df.loc[get_path_geo, 'filepaths'].values[0]
print(file_path_geo)

c:\Users\guicr\OneDrive\Carreira\Portfolio\uber\data\sp_setores_censitarios\35SEE250GC_SIR.shp


Utiliza-se a biblioteca geopandas para tratamento dos dados geográficos.

In [76]:
cities_setores_censitarios = gpd.read_file(file_path_geo)

In [77]:
cities_setores_censitarios.head().T

0  \
ID                                                    98237.0   
CD_GEOCODI                                    354100005000009   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         1   
geometry    POLYGON ((-46.410239 -24.011028, -46.411694999...   

                                                            1  \
ID                                                    98232.0   
CD_GEOCODI                                    354100005000004   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         2   
geometry    POLYGON ((-46.416055 -24.002446, -46.415983999...   

                                                            2  \
ID                                                    98230.0   
CD_GEOCODI                                    354100005000002   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         3   
geometry    POLYGON ((-46.412132 -24.007921, -46.412442 -2...   

                                                            3  \
ID                                                    98229.0   
CD_GEOCODI                                    354100005000001   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                        

In [78]:
# cópia do arquivo 'cru'
stg_cities_setores_censitarios = cities_setores_censitarios.copy()

##### Verificando os dados...

In [79]:
stg_cities_setores_censitarios.head(10).T

0  \
ID                                                    98237.0   
CD_GEOCODI                                    354100005000009   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         1   
geometry    POLYGON ((-46.410239 -24.011028, -46.411694999...   

                                                            1  \
ID                                                    98232.0   
CD_GEOCODI                                    354100005000004   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         2   
geometry    POLYGON ((-46.416055 -24.002446, -46.415983999...   

                                                            2  \
ID                                                    98230.0   
CD_GEOCODI                                    354100005000002   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         3   
geometry    POLYGON ((-46.412132 -24.007921, -46.412442 -2...   

                                                            3  \
ID                                                    98229.0   
CD_GEOCODI                                    354100005000001   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                        

Os dados geográficos apresentam as seguintes relevantes para essa análise:
1. CD_GEOCODI - Este é código do setor censitário para referenciarmos ao censo 2010;
2. geometry - Dados do polígono que contem o censo censitário para cálculo da área do mesmo.

In [80]:
stg_cities_setores_censitarios.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 68296 entries, 0 to 68295
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   ID          68296 non-null  float64 
 1   CD_GEOCODI  68296 non-null  object  
 2   TIPO        68296 non-null  object  
 3   CD_GEOCODS  68296 non-null  object  
 4   NM_SUBDIST  0 non-null      float64 
 5   CD_GEOCODD  68296 non-null  object  
 6   NM_DISTRIT  68296 non-null  object  
 7   CD_GEOCODM  68296 non-null  object  
 8   NM_MUNICIP  68296 non-null  object  
 9   NM_MICRO    68296 non-null  object  
 10  NM_MESO     68296 non-null  object  
 11  CD_GEOCODB  18797 non-null  object  
 12  NM_BAIRRO   18797 non-null  object  
 13  ID1         68296 non-null  int64   
 14  geometry    68296 non-null  geometry
dtypes: float64(2), geometry(1), int64(1), object(11)
memory usage: 7.8+ MB


In [81]:
stg_cities_setores_censitarios.isnull().sum()

ID                0
CD_GEOCODI        0
TIPO              0
CD_GEOCODS        0
NM_SUBDIST    68296
CD_GEOCODD        0
NM_DISTRIT        0
CD_GEOCODM        0
NM_MUNICIP        0
NM_MICRO          0
NM_MESO           0
CD_GEOCODB    49499
NM_BAIRRO     49499
ID1               0
geometry          0
dtype: int64

Ao analisar as informações percebe-se que todos os dados necessários estão completos sem valores faltantes.

In [82]:
# Renomeando a coluna 'CD_GEOCODI' para 'Cod_setor' a fim de criar a mesma referência em ambos os dataframes
stg_cities_setores_censitarios.rename(columns={'CD_GEOCODI':'Cod_setor'}, inplace=True)
stg_cities_setores_censitarios.head().T

0  \
ID                                                    98237.0   
Cod_setor                                     354100005000009   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         1   
geometry    POLYGON ((-46.410239 -24.011028, -46.411694999...   

                                                            1  \
ID                                                    98232.0   
Cod_setor                                     354100005000004   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         2   
geometry    POLYGON ((-46.416055 -24.002446, -46.415983999...   

                                                            2  \
ID                                                    98230.0   
Cod_setor                                     354100005000002   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         3   
geometry    POLYGON ((-46.412132 -24.007921, -46.412442 -2...   

                                                            3  \
ID                                                    98229.0   
Cod_setor                                     354100005000001   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                        

O cálculo de área é realizado na projeção UTM, portanto será necessário realizar a mudança de coordenada geográfica para cartesiana.

In [83]:
# Mudança de coordenada
stg_cities_setores_censitarios = stg_cities_setores_censitarios.to_crs(epsg=31983) # Coordenada UTM para São Paulo é a EPSG 31983
stg_cities_setores_censitarios['geometry'].head()

0    POLYGON ((356565.548 7343834.128, 356417.318 7...
1    POLYGON ((355964.368 7344778.533, 355971.068 7...
2    POLYGON ((356369.535 7344176.257, 356337.978 7...
3    POLYGON ((356447.233 7344442.942, 356383.829 7...
4    POLYGON ((356251.286 7344483.503, 356090.929 7...
Name: geometry, dtype: geometry

In [84]:
# Cálculo da área para hectares
stg_cities_setores_censitarios['area_ha'] = stg_cities_setores_censitarios['geometry'].area / 1e4

O retorno para a coordenada geográfica é necessária para a representação visual em mapas posteriormente.

In [85]:
# Retorno para coordenada geográfica
stg_cities_setores_censitarios = stg_cities_setores_censitarios.to_crs(epsg=4326)

In [86]:
stg_cities_setores_censitarios.head().T

0  \
ID                                                    98237.0   
Cod_setor                                     354100005000009   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         1   
geometry    POLYGON ((-46.410239 -24.011028, -46.411695 -2...   
area_ha                                              2.382885   

                                                            1  \
ID                                                    98232.0   
Cod_setor                                     354100005000004   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         2   
geometry    POLYGON ((-46.416055 -24.002445999999996, -46....   
area_ha                                             10.134651   

                                                            2  \
ID                                                    98230.0   
Cod_setor                                     354100005000002   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                       354100005001   
NM_BAIRRO                                           Boqueirão   
ID1                                                         3   
geometry    POLYGON ((-46.412132 -24.007920999999996, -46....   
area_ha                                              3.067219   

                                                            3  \
ID                                                    98229.0   
Cod_setor                                     354100005000001   
TIPO                                                   URBANO   
CD_GEOCODS                                        35410000500   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          354100005   
NM_DISTRIT                                       PRAIA GRANDE   
CD_GEOCODM                                            3541000   
NM_MUNICIP                                       PRAIA GRANDE   
NM_MICRO                                               SANTOS   
NM_MESO                    

In [87]:
# Transformando a coluna 'Cod_setor' para string para posterior mesclagem dos dataframes 
stg_cities_setores_censitarios['Cod_setor'] = stg_cities_setores_censitarios['Cod_setor'].astype('str')

    Consolidação dos dados geográficos com os dados dos domicílios.

In [100]:
stg_cities_setores_censitarios.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 68296 entries, 0 to 68295
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   ID          68296 non-null  float64 
 1   Cod_setor   68296 non-null  object  
 2   TIPO        68296 non-null  object  
 3   CD_GEOCODS  68296 non-null  object  
 4   NM_SUBDIST  0 non-null      float64 
 5   CD_GEOCODD  68296 non-null  object  
 6   NM_DISTRIT  68296 non-null  object  
 7   CD_GEOCODM  68296 non-null  object  
 8   NM_MUNICIP  68296 non-null  object  
 9   NM_MICRO    68296 non-null  object  
 10  NM_MESO     68296 non-null  object  
 11  CD_GEOCODB  18797 non-null  object  
 12  NM_BAIRRO   18797 non-null  object  
 13  ID1         68296 non-null  int64   
 14  geometry    68296 non-null  geometry
 15  area_ha     68296 non-null  float64 
dtypes: float64(3), geometry(1), int64(1), object(11)
memory usage: 8.3+ MB


A dimensionalidade dos dados geoográficos, também, é maior do que os dados dos domicílios. Portanto, mesclam-se os dataframes levando isso em conta.

In [101]:
comb1_cities = pd.merge(comb_city, stg_cities_setores_censitarios, how='left', on='Cod_setor')

In [102]:
comb1_cities.head().T

0  \
Cod_setor                                     355030801000001   
renda                                                907777.0   
domicilios                                                268   
ID                                                   108939.0   
TIPO                                                   URBANO   
CD_GEOCODS                                        35503080100   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          355030801   
NM_DISTRIT                                          ÁGUA RASA   
CD_GEOCODM                                            3550308   
NM_MUNICIP                                          SÃO PAULO   
NM_MICRO                                            SÃO PAULO   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                                NaN   
NM_BAIRRO                                                 NaN   
ID1                                                     43401   
geometry    POLYGON ((-46.569514 -23.569162999999996, -46....   
area_ha                                              7.213308   

                                                            1  \
Cod_setor                                     355030801000002   
renda                                                846525.0   
domicilios                                                306   
ID                                                   108940.0   
TIPO                                                   URBANO   
CD_GEOCODS                                        35503080100   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          355030801   
NM_DISTRIT                                          ÁGUA RASA   
CD_GEOCODM                                            3550308   
NM_MUNICIP                                          SÃO PAULO   
NM_MICRO                                            SÃO PAULO   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                                NaN   
NM_BAIRRO                                                 NaN   
ID1                                                     43402   
geometry    POLYGON ((-46.568704 -23.568531999999994, -46....   
area_ha                                              7.162393   

                                                            2  \
Cod_setor                                     355030801000003   
renda                                                505662.0   
domicilios                                                189   
ID                                                   108941.0   
TIPO                                                   URBANO   
CD_GEOCODS                                        35503080100   
NM_SUBDIST                                                NaN   
CD_GEOCODD                                          355030801   
NM_DISTRIT                                          ÁGUA RASA   
CD_GEOCODM                                            3550308   
NM_MUNICIP                                          SÃO PAULO   
NM_MICRO                                            SÃO PAULO   
NM_MESO                            METROPOLITANA DE SÃO PAULO   
CD_GEOCODB                                                NaN   
NM_BAIRRO                                                 NaN   
ID1                                                     43403   
geometry    POLYGON ((-46.568557 -23.56965199999999, -46.5...   
area_ha                                              5.596828   

                                                            3  \
Cod_setor                                     355030801000004   
renda                                                446011.0   
domicilios                                                181   
ID                                                   108942.0   
TIPO                       

In [103]:
comb1_cities.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64717 entries, 0 to 64716
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   Cod_setor   64717 non-null  object  
 1   renda       64717 non-null  float64 
 2   domicilios  64717 non-null  int32   
 3   ID          64717 non-null  float64 
 4   TIPO        64717 non-null  object  
 5   CD_GEOCODS  64717 non-null  object  
 6   NM_SUBDIST  0 non-null      float64 
 7   CD_GEOCODD  64717 non-null  object  
 8   NM_DISTRIT  64717 non-null  object  
 9   CD_GEOCODM  64717 non-null  object  
 10  NM_MUNICIP  64717 non-null  object  
 11  NM_MICRO    64717 non-null  object  
 12  NM_MESO     64717 non-null  object  
 13  CD_GEOCODB  17838 non-null  object  
 14  NM_BAIRRO   17838 non-null  object  
 15  ID1         64717 non-null  int64   
 16  geometry    64717 non-null  geometry
 17  area_ha     64717 non-null  float64 
dtypes: float64(4), geometry(1), int32(1), int64(1)

---

## EDA - Análise exploratória de dados

In [ ]:
len(stg_cities_setores_censitarios['Cod_setor'])

In [ ]:
len(stg1_city_all['Cod_setor'])

Verifica-se que existem maior quantidade de setores censitários no dataframe geográfico do que no dataframe dos dados do censo. Isso decorre da legislação brasileira, que exige que as instituições responsáveis por realizar pesquisas estatísticas forneçam informações ao IBGE e protejam os dados de forma a garantir a não identificação dos informantes. Isso significa que, para proteger os dados, o IBGE pode usar medidas como restrição de dados (reduzir o volume de informação liberada) ou restrição de acesso (estabelecer condições para o uso dos dados). No caso do Censo Demográfico 2010, o IBGE optou por restrição de dados e omitiu os valores das maioria das variáveis em setores com menos de cinco domicílios particulares permanentes, conforme a documentação.

Realiza-se a mesclagem dos dados tendo como base os dados do censo para continuação do EDA.

In [ ]:
# Copiando apenas os dados relevantes para o EDA
stg1_cities_setores_censitarios = stg_cities_setores_censitarios[['Cod_setor', 'geometry', 'area_ha']].copy()

In [ ]:
stg1_cities_setores_censitarios.head().T

In [ ]:
# Mesclando todos os dados, nesta  mesclagem fica-se apenas com os dados do censo
stg_cities = pd.merge(stg1_city_all, stg1_cities_setores_censitarios, on='Cod_setor')

In [ ]:
stg_cities.head().T

In [ ]:
stg_cities.info()

In [ ]:
stg_cities.isnull().sum()

Todas as informações foram mescladas.

Cálculo da densidade demográfica como um índice de aglomeração para entendimento do adensamento populacional.

In [ ]:
stg_cities['dens_pop'] = stg_cities['pop'] / stg_cities['area_ha']

In [ ]:
stg_cities.head().T

Neste momento, analisa-se algumas informações relevantes sobre os municípios. Os dois aspectos a serem observados serão a renda e densidade populacional para cumprimento do objetivo desta análise.

In [ ]:
municipios_gb =  stg_cities.groupby('Nome_do_municipio').agg({'renda':'mean', 'pop':'sum', 'area_ha':'sum'})
municipios_gb['dens_pop'] = municipios_gb['pop'] / municipios_gb['area_ha']

In [ ]:
dens_mun = municipios_gb.sort_values(by='dens_pop', ascending=False)
dens_mun

In [ ]:
renda_mun = municipios_gb.sort_values(by='renda', ascending=False)
renda_mun

Quais são os top 10?

In [ ]:
dens_mun[:10]

In [ ]:
renda_mun[:10]

Assim, conseguimos saber para os municípios de São Paulo quais possuem maior renda e densidade demográfica. No entanto, o que agrega valor ao estudo é saber, por setores censitários de cada município, onde estão as maiores rendas e densidades. 

De acordo com as autoras do artigo "NÍVEIS DE DENSIDADE POPULACIONAL: Uma proposta de classificação para Goiânia – GO, aplicação no Setor Central" (https://xviiienanpur.anpur.org.br/anaisadmin/capapdf.php?reqid=1034) podemos classificar as densidades populacionais como, em habitantes/hectare, " <15 muito baixa, de 16 a 50 baixa, de 51 a 150 média, de 151 a 350 alta e > 350 muito alta."

Além disso, utilizando a informação do artigo do site einvestidor.estadao.com.br ("Quanto ganha a classe A no Brasil?" Quanto ganha a classe A no Brasil? https://einvestidor.estadao.com.br/radar-einvestidor/quanto-ganha-a-classe-a-no-brasil/) obtém-se os dados referentes as rendas domiciliares por classe social:
- Classe A: R$ 28.240;
- Classe B1: R$ 12.683,34;
- Classe B2: R$ 7.017,64;
- Classe C1: R$ 3.980,38;
- Classe C2: R$ 2.403,04;
- Classe D-E: R$ 1.087,77.

Ademais, o artigo Tendências e Desigualdades da Mobilidade Urbana no Brasil II: Características e Padrões de Consumo da Mobilidade por Aplicativo”, conduzido por Lucas Warwar e Rafael H. M. Pereira (https://repositorio.ipea.gov.br/bitstream/11058/11240/1/td_2781.pdf) demonstra que "A distribuição dos usuários de ride-hailing (viagens por aplicativos) por faixas de renda evidencia uma forte concentração entre os estratos superiores".

De acordo com informações anteriores, consegue-se obter os dados relevantes das condições sócio-demográficas para a descoberta das regiões mais propensas a se obter viagens por aplicativo. Portanto, os dados deverão ser estratificados em regiões com densidade demográfica maior que 150 hab/ha e renda domiciliar acima de R$ 7.017,64. No entanto, os dados tratados, até este momento, não possuem as informações por domícilio, porém os dados do IBGE possuem tal valor. Assim, outro arquivo será criado para tratamento destes dados.

---

### Preparação da visualização a ser utilizada no arquivo atualizado.

Visualização de dados

In [ ]:
plt.figure(figsize=(15,5))

plt.subplot(1,2,1)
plt.hist(stg_city['renda'], bins=30, edgecolor='k')
plt.xlabel('R$ per capita')
plt.ylabel('Frequencia')

plt.subplot(1,2,2)
stg_city.boxplot(column='renda')
plt.ylabel('R$ per capita')

plt.show()

In [ ]:
plt.figure(figsize=(15,5))

plt.subplot(1,2,1)
plt.hist(stg_city['pop'], bins=30, edgecolor='k')
plt.xlabel('População - hab')
plt.ylabel('Frequencia')

plt.subplot(1,2,2)
stg_city.boxplot(column='pop')
plt.ylabel('População')

plt.show()

In [ ]:
plt.figure(figsize=(15,5))

plt.subplot(1,2,1)
plt.hist(stg_city['dens_pop'], bins=30, edgecolor='k')
plt.xlabel('Densidade populacional - hab/ha')
plt.ylabel('Frequencia')

plt.subplot(1,2,2)
stg_city.boxplot(column='dens_pop')
plt.ylabel('Densidade populacional - hab/ha')

plt.show()

In [ ]:
stg_city.describe()

In [ ]:
q3_renda = stg_city['renda'].quantile(0.75)
q3_dens_pop = stg_city['dens_pop'].quantile(0.50)

mask = (stg_city['renda'] >= q3_renda) & (stg_city['dens_pop'] >= q3_dens_pop)
top_25_stg_city = stg_city[mask].copy()
top_25_stg_city.head().T

In [ ]:
top_25_stg_city.info()

In [ ]:
top_25_stg_city.describe()

Visualização geográfica

In [ ]:
top_25_stg_city_gdf = gpd.GeoDataFrame(top_25_stg_city, geometry='geometry')

In [ ]:
hotels_list_path = os.path.join(data_dir, 'hotels_list.csv')

hotels_list_df = pd.read_csv(hotels_list_path)

hotels_list_df

In [ ]:
map = folium.Map(location=[-23.5633, -46.66744],
                 tiles='Cartodb Positron',
                 zoom_start=13)

borders_style = {
    'color': 'green',
    'weight': 1,
    'fillColor': 'green',
    'fillOpacity': 0.15,
}

top_25 = folium.GeoJson(data=top_25_stg_city_gdf,
                             name='Top 25% renda e densidade populacional',
                             style_function=lambda x: borders_style,
                             )

top_25.add_to(map)

for index, row in hotels_list_df.iterrows():
    folium.Marker(
        location=[row['latitude'], row['longitude']],
        popup=row['name'],
    ).add_to(map)

map_dir = os.path.join(artifacts_dir, "top_25_sao_paulo.html")

map.save(map_dir)